realm/realm-library/src/main/java/io/realm/RealmCache.java

```diff
@@ -0,0 +1,269 @@
+/*
+ * Copyright 2015 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.realm;
+
+import java.util.Arrays;
+import java.util.EnumMap;
+import java.util.HashMap;
+import java.util.Map;
+
+import io.realm.internal.ColumnIndices;
+import io.realm.internal.RealmProxyMediator;
+import io.realm.internal.SharedGroup;
+import io.realm.internal.log.RealmLog;
+
+/**
+ * To cache {@link Realm}, {@link DynamicRealm} instances and related resources.
+ * Every thread will share the same {@link Realm} and {@link DynamicRealm} instances which are referred to the same
+ * {@link RealmConfiguration}.
+ */
+class RealmCache {
+
+    private class RefAndCount {
+        // The Realm instance in this thread.
+        private final ThreadLocal<BaseRealm> localRealm = new ThreadLocal<>();
+        // How many references to this Realm instance in this thread.
+        private final ThreadLocal<Integer> localCount = new ThreadLocal<>();
+        // How many threads have instances refer to this configuration.
+        private int globalCount = 0;
```

Comment:
Mixing ThreadLocal and class variables in this class seems really dangerous to me? It would only be safe if you wrap any access to globalCount in a lock (granted that seems to be happening in the code below, but still). It would feel much safer if you could only access the globalCount through a setter in the class which could guarantee that no race conditions happened? Converting it to an AtomicInt would also solve it?